src/sidebar/store/create-store.js

```diff
@@ -83,6 +81,74 @@ import { createReducer, bindSelectors } from './util';
  *   SelectorMethods<RootSelectors>} Store
  */
 
+/**
+ * Create a Redux reducer from a store module's reducer map.
+ *
+ * @template State
+ * @param {ReducerMap<State>} reducers
+ */
+function createReducer(reducers) {
+  /** @param {redux.Action} action */
+  return (state = /** @type {State} */ ({}), action) => {
+    const reducer = reducers[action.type];
+    if (!reducer) {
+      return state;
+    }
+    const stateChanges = reducer(state, action);
+
+    // Some modules return an array rather than an object. They need to be
+    // handled differently so we don't convert them to an object.
+    if (Array.isArray(stateChanges)) {
+      return stateChanges;
+    }
+
+    return {
+      ...state,
+      ...stateChanges,
+    };
+  };
+}
+
+/**
+ * Convert a map of selector functions, which take a state value as their
+ * first argument, to a map of selector methods, which pre-fill the first
+ * argument by calling `getState()`.
+ *
+ * @template State
+ * @template {SelectorMap<State>} Selectors
+ * @param {Selectors} selectors
+ * @param {() => State} getState
+ * @return {SelectorMethods<Selectors>}
+ */
+function bindSelectors(selectors, getState) {
+  /** @type {Record<string, Function>} */
+  const boundSelectors = {};
+  for (let [name, selector] of Object.entries(selectors)) {
+    boundSelectors[name] = /** @param {any[]} args */ (...args) =>
+      selector(getState(), ...args);
+  }
+  return /** @type {SelectorMethods<Selectors>} */ (boundSelectors);
+}
+
+/**
+ * `Object.assign` wrapper that checks for overwriting properties in debug builds.
+ *
+ * @template T
+ * @template U
+ * @param {T} target
+ * @param {U} source
```

Comment:
Checking my understanding and learning comprehension: Using `@template` here because inputs (params) to this function are relevant in inferring the return value?

Reply:
Yes, indeed. If you hover the `assignOnce` definition in your editor, you'll see that the return type is inferred as `T & U`. This comes from how `Object.assign` is defined (`Object.assign(T, U)` returns `T & U`).